### PR TITLE
Add search-autocomplete UI component

### PR DESCRIPTION
**Pull request by IgorShmelev**
This PR adds an input field with planned search and autocomplete functionality for nodes/edges/entities, visible on the subpages of the app. 
The functionality for retrieving the information needs to be extended with the fitting service.

### CODE REVIEW

File: frontend/package.json [hunk 1]
```diff
@@ -27,6 +27,7 @@
     "@cypress/react": "^5.4.0",
     "@material-ui/core": "^4.11.4",
     "@material-ui/icons": "^4.11.2",
+    "@material-ui/lab": "^4.0.0-alpha.58",
```
Comment (CatoLeanTruetschel) [inline frontend/package.json:30]: This is non-stable technology. Why is this needed?

File: frontend/src/Layout.tsx [hunk 2]
```diff
@@ -172,6 +180,18 @@ function Layout(props: LayoutProps): JSX.Element {
     );
   }
 
+  function RenderSearchbar(): JSX.Element | null {
+    if (currentPath === '/home') {
```
Comment (CatoLeanTruetschel) [inline frontend/src/Layout.tsx:184]: Why is there no search on the startpage?

File: frontend/src/search/SearchResultOption.ts [hunk 3]
```diff
@@ -0,0 +1,9 @@
+/**
+ * Represents a option element inside the dropdown of the autocomplete search.
+ */
+export interface SearchResultOption {
```
Comment (CatoLeanTruetschel) [inline frontend/src/search/SearchResultOption.ts:4]: Rename to `SearchResultsEntry`

File: frontend/src/search/SearchResultOption.ts [hunk 4]
```diff
@@ -0,0 +1,9 @@
...
+  label: string,
+  subLabel: string,
+  type: string,
+  color?: string | undefined
+}
```
Comment (CatoLeanTruetschel) [inline frontend/src/search/SearchResultOption.ts:9]: No newline

**Review by CatoLeanTruetschel — CHANGES_REQUESTED** [review]

File: frontend/src/Layout.tsx (hunk 2 above)
Comment (IgorShmelev) [reply frontend/src/Layout.tsx:184]: Based on the user story #30 the search field should be displayed on every subpage. In my opinion, this affects all pages after the app entry page - Home.

File: frontend/src/search/Searchbar.tsx [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+
+          if (value) {
+            newOptions = [value];
+          }
+
+          if (resultOptions) {
+            newOptions = [...newOptions, ...resultOptions];
+          }
+
+          setOptions(newOptions);
```
Comment (joluj) [inline frontend/src/search/Searchbar.tsx:66]: I think this won't be executed in the `useEffect` runcycle since it's an async operation

**Review by joluj — CHANGES_REQUESTED** [review]

File: frontend/src/Layout.tsx (hunk 2 above)
Comment (CatoLeanTruetschel) [reply frontend/src/Layout.tsx:184]: No, this just states that there should not be a dedicated search page, but the search shall be accessible from every page.

File: frontend/src/search/SearchResultEntry.ts [hunk 6]
```diff
@@ -0,0 +1,10 @@
+/**
+ * Represents a option element inside the dropdown of the autocomplete search.
+ */
+export interface SearchResultEntry {
```
Comment (joluj) [inline frontend/src/search/SearchResultEntry.ts:4]: Add a comment to every (/some) attributes. I don't know what `subLabel` should be.

File: frontend/src/search/Searchbar.tsx [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+  const searchService = useService(SearchService, null);
+
+  const loadOptions = function (
+    inputValue: string,
+    callback: (entries: SearchResultEntry[]) => void
+  ) {
+    searchService
+      .fullTextSearch(inputValue)
+      .then((result) => callback(convertSearchResultToEntries(result)));
+    // TODO Error handling for bad requests
```
Comment (joluj) [inline frontend/src/search/Searchbar.tsx:25]: #128

File: frontend/src/search/Searchbar.tsx [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+  };
+
+  /* eslint-disable react/jsx-no-bind */
```
Comment (joluj) [inline frontend/src/search/Searchbar.tsx:28]: Add a comment why you disabled that.
`/* eslint-disable react/jsx-no-bind -- my cool comment */`

**Review by joluj — CHANGES_REQUESTED** [review]
Looks promising and very nice!
Opened #128 and #129.
Tests still missing.

File: frontend/src/search/SearchResultEntryGroup.ts [hunk 9]
```diff
@@ -0,0 +1,17 @@
+import { SearchResultEntry } from './SearchResultEntry';
+
+/**
+ * Represents a group of option elements inside the dropdown of the autocomplete search.
+ * The groups are separated by the type of search results (node, edge, node-type, edge-type)
+ */
+
```
Comment (joluj) [inline frontend/src/search/SearchResultEntryGroup.ts:7]:
```suggestion

```

File: frontend/src/search/SearchResultEntryGroup.ts [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+export interface SearchResultEntryGroup {
+  /**
+   * group label that is visible to the user
+   */
+  label: string,
```
Comment (joluj) [inline frontend/src/search/SearchResultEntryGroup.ts:12]:
```suggestion
  label: string;
```

Use the linter (`yarn lint --fix` might even do it)

File: frontend/src/search/Searchbar.tsx [hunk 11]
```diff
@@ -0,0 +1,38 @@
+import React from 'react';
+import AsyncSelect from 'react-select/async';
+import useService from '../dependency-injection/useService';
+import SearchService from '../services/searchService';
+import { SearchResultEntryGroup } from './SearchResultEntryGroup';
+import convertSearchResultToEntryGroups from './SearchEntryConverter';
+
+export default function Searchbar(): JSX.Element {
+  /* eslint-disable-next-line  @typescript-eslint/no-explicit-any -- no use for the explicit type  */
```
Comment (joluj) [inline frontend/src/search/Searchbar.tsx:9]:
```suggestion

```

> no use for the explicit type

Does type `unknown` also work?

File: frontend/src/search/Searchbar.tsx [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+  const searchService = useService(SearchService, null);
+
+  const searchEntries = function executeSearch(
+    inputValue: string,
+    callback: (entryGroups: SearchResultEntryGroup[]) => void
+  ) {
+    searchService
+      .fullTextSearch(inputValue)
+      .then((result) => callback(convertSearchResultToEntryGroups(result)))
+      // TODO what can we really do with the error here?
```
Comment (joluj) [inline frontend/src/search/Searchbar.tsx:25]: #112

**Review by joluj — CHANGES_REQUESTED** [review]
#128 #129 Still open

**Review by CatoLeanTruetschel — APPROVED** [review]
